File: src/HTML/extensions/sip/ov_sip.js
/***
    ---------------------------------------------------------------------------

    Copyright (c) 2018 German Aerospace Center DLR e.V. (GSOC)

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

            http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.

    This file is part of the openvocs project. https://openvocs.org

    ---------------------------------------------------------------------------
*//**
    @file           ov_sip.js

    @ingroup        extensions/sip

    @brief          implements openvocs sip protocol and wraps it
                    around a list of ov_websockets

    ---------------------------------------------------------------------------
*/

import * as ov_Websockets from "/lib/ov_websocket_list.js";

export const EVENT = {
    SIP_CALL: "sip_call",
    SIP_HANGUP: "sip_hangup",
    SIP_PERMIT: "sip_permit_call",
    SIP_REVOKE: "sip_revoke_call",
    SIP_LIST_CALLS: "sip_get_calls",
    SIP_LIST_CALL_PERMISSIONS: "sip_get_call_permissions",
    SIP_LIST_STATUS: "sip_get_status"
};

// SIP --------------------------------------------------------------------
export async function sip(websocket) {
    if (websocket)
        return await ws_sip(websocket);

    let lead_promise;
    for (let ws of ov_Websockets.list) {
        if (ws.authenticated) {
            let promise = ws_sip(ws);
            if (ws === ov_Websockets.prime_websocket)
                lead_promise = promise;
        }
    }
    return await lead_promise;
}
export async function sip_call(loop_id, from, to, websocket) {
    if (websocket)
        return await ws_sip_call(loop_id, from, to, websocket);

    let lead_promise;
    for (let ws of ov_Websockets.list) {
        if (ws.authenticated) {
            let promise = ws_sip_call(loop_id, from, to, ws);
            if (ws === ov_Websockets.prime_websocket)
                lead_promise = promise;
        }
    }
    return await lead_promise;
}

export async function sip_hangup(loop_id, call_id, websocket) {
    if (websocket)
        return await ws_sip_hangup(loop_id, call_id, websocket);

    let lead_promise;
    for (let ws of ov_Websockets.list) {
        if (ws.authenticated) {
            let promise = ws_sip_call(call_id, ws);
            if (ws === ov_Websockets.prime_websocket)
                lead_promise = promise;
        }
    }
    return await lead_promise;
}

export async function sip_list_calls(websocket) {
    if (websocket)
        return await ws_sip_list_calls(websocket);

    let lead_promise;
    for (let ws of ov_Websockets.list) {
        if (ws.authenticated) {
            let promise = ws_sip_list_calls(ws);
            if (ws === ov_Websockets.prime_websocket)
                lead_promise = promise;
        }
    }
    return await lead_promise;
}

export async function sip_permit(loop_id, caller, callee, websocket) {
    if (websocket)
        return await ws_sip_permit(loop_id, caller, callee, websocket);

    let lead_promise;
    for (let ws of ov_Websockets.list) {
        if (ws.authenticated) {
            let promise = ws_sip_permit(loop_id, caller, callee, websocket);
            if (ws === ov_Websockets.prime_websocket)
                lead_promise = promise;
        }
    }
    return await lead_promise;
}

export async function sip_revoke(loop_id, caller, callee, websocket) {
    if (websocket)
        return await ws_sip_revoke(loop_id, caller, callee, websocket);

    let lead_promise;
    for (let ws of ov_Websockets.list) {
        if (ws.authenticated) {
            let promise = ws_sip_revoke(loop_id, caller, callee, websocket);
            if (ws === ov_Websockets.prime_websocket)
                lead_promise = promise;
        }
    }
    return await lead_promise;
}

/*list_permissions() {
    return this.#request(this.#create_event(EVENT.SIP_LIST_CALL_PERMISSIONS));
}

list_status() {
    return this.#request(this.#create_event(EVENT.SIP_LIST_STATUS));
}*/

async function ws_sip_call(loop_id, from, to, websocket) {
    try {
        console.log(log_prefix(websocket) + "requesting sip connection from loop " + loop_id + " to " + to + "...");
        let parameter = { loop: loop_id, destination: to, from: from };
        let result = await websocket.send_event(EVENT.SIP_CALL, parameter);
        console.log(log_prefix(websocket) + "waiting for sip to connect...");
        return result;
    } catch (error) {
        console.warn(log_prefix(websocket) + "requesting sip connection failed.", error.error);
        return false;
    }
}

async function ws_sip_hangup(loop_id, call_id, websocket) {
    try {
        console.log(log_prefix(websocket) + "requesting sip hangup for call " + call_id + "...");
        let parameter = { call: call_id, loop: loop_id };
        let result = await websocket.send_event(EVENT.SIP_HANGUP, parameter);
        console.log(log_prefix(websocket) + "call " + call_id + " ended");
        return result;
    } catch (error) {
        console.warn(log_prefix(websocket) + "requesting sip connection failed.", error);
        return false;
    }
}

async function ws_sip_list_calls(websocket) {
    try {
        console.log(log_prefix(websocket) + "requesting current list of sip calls...");
        let result = await websocket.send_event(EVENT.SIP_LIST_CALLS);
        console.log(log_prefix(websocket) + "received list of sip calls");
        return result.calls;
    } catch (error) {
        console.warn(log_prefix(websocket) + "requesting list of sip calls failed.", error);
        return false;
    }
}

async function ws_sip_permit(loop_id, caller, callee, websocket) {
    try {
        console.log(log_prefix(websocket) + "add sip permit...");
        let parameter = { loop: loop_id, caller: caller, callee: callee };
        let result = await websocket.send_event(EVENT.SIP_PERMIT, parameter);
        console.log(log_prefix(websocket) + "added sip permit");
        return result;
    } catch (error) {
        console.warn(log_prefix(websocket) + "adding sip permit failed.", error);
        return false;
    }
}

async function ws_sip_revoke(loop_id, caller, callee, websocket) {
    try {
        console.log(log_prefix(websocket) + "revoke sip permit...");
        let parameter = { loop: loop_id, caller: caller, callee: callee };
        let result = await websocket.send_event(EVENT.SIP_REVOKE, parameter);
        console.log(log_prefix(websocket) + "revoked sip permit");
        return result;
    } catch (error) {
        console.warn(log_prefix(websocket) + "revoking sip permit failed.", error);
        return false;
    }
}

//-----------------------------------------------------------------------------
// logging
//-----------------------------------------------------------------------------
function log_prefix(ws) {
    return "(" + ws.server_name + " sip) ";
}

File: src/samples/ov_sip/src/ov_sip.c
/***
        ------------------------------------------------------------------------

        Copyright (c) 2025 German Aerospace Center DLR e.V. (GSOC)

        Licensed under the Apache License, Version 2.0 (the "License");
        you may not use this file except in compliance with the License.
        You may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

        Unless required by applicable law or agreed to in writing, software
        distributed under the License is distributed on an "AS IS" BASIS,
        WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
        See the License for the specific language governing permissions and
        limitations under the License.

        This file is part of the openvocs project. https://openvocs.org

        ------------------------------------------------------------------------
*//**
        @author         Michael J. Beer

        ------------------------------------------------------------------------
*/
#include <ov_base/ov_socket.h>
#include <ov_base/ov_utils.h>
#include <ov_os/ov_os_event_loop.h>
#include <ov_sip/ov_sip_app.h>
#include <stdarg.h>
#include <stdio.h>

/*----------------------------------------------------------------------------*/

FILE *outstream = 0;

/*----------------------------------------------------------------------------*/

static void print(char const *msg, ...) {

  va_list args;
  va_start(args, msg);

  fprintf(outstream, "%lld: ", (long long)time(0));
  vfprintf(outstream, msg, args);

  va_end(args);
}

/*----------------------------------------------------------------------------*/

void cb_closed(int fd, void *additional) {

  UNUSED(fd);

  char **constringptr = additional;
  char *constring = *constringptr;
  print("Connection %s closed\n", constring);

  constring = ov_free(constring);
  constring = strdup("UNCONNECTED");
  *constringptr = constring;
}

/*----------------------------------------------------------------------------*/

char *constring_from_fd(int fd) {

  ov_socket_state state = {0};

  if (ov_socket_state_from_handle(fd, &state)) {
    //              "Peer:"  hostname  + ":"  + port number + '\0'
    size_t bufsize = 5 + OV_HOST_NAME_MAX + 1 + 6 + 1;
    char *con = calloc(1, bufsize);
    snprintf(con, bufsize, "Peer:%s:%" PRIu16, state.peer.host,
             state.peer.port);
    return con;

  } else {
    print("Could not get socket state from handle\n");
    return strdup("New connection");
  }
}

/*----------------------------------------------------------------------------*/

void cb_reconnected(int fd, void *additional) {
  char **constringptr = additional;
  char *constring = constring_from_fd(fd);
  print("Reconnected successfully %s\n", constring);

  ov_free(*constringptr);
  *constringptr = constring;
}

/*----------------------------------------------------------------------------*/

void cb_accepted(int fd, void *additional) {

  char **constringptr = additional;
  char *constring = constring_from_fd(fd);
  print("Accepted new connection %s - beware: We assume only ONE client "
        "connection - if there are more, the logging will be wrong\n",
        constring);

  ov_free(*constringptr);
  *constringptr = constring;
}

/*****************************************************************************
                                  SIP HANDLERS
 ****************************************************************************/

static void cb_sip_response(ov_sip_message const *message, int fd,
                            void *additional) {
  UNUSED(fd);
  UNUSED(message);

  char **constringptr = additional;

  print("Received new SIP response from %s\n", *constringptr);
  ov_sip_message_dump(outstream, message);
  fprintf(outstream,
          "------------------------------------------------------------\n");
}

/*----------------------------------------------------------------------------*/

static void cb_sip_request(ov_sip_message const *message, int fd,
                           void *additional) {
  UNUSED(fd);
  UNUSED(message);

  char **constringptr = additional;

  print("Received new SIP request from %s:\n", *constringptr);
  ov_sip_message_dump(outstream, message);
  fprintf(outstream,
          "------------------------------------------------------------\n");
}

/*----------------------------------------------------------------------------*/

int main(int argc, char *argv[]) {
  UNUSED(argc);
  UNUSED(argv);

  outstream = stderr;

  ov_socket_configuration server_socket_cfg = {
      .host = "127.0.0.1",
      .port = 1111,
      .type = TCP,
  };

  char *constring = strdup("UNCONNECTED");

  ov_sip_app_configuration cfg = {

      .log_io = true,
      .reconnect_interval_secs = 20,
      .accept_to_io_timeout_secs = 20,

      .cb_closed = cb_closed,
      .cb_reconnected = cb_reconnected,
      .cb_accepted = cb_accepted,

      .additional = &constring,
      .are_methods_case_sensitive = false,

  };

  ov_event_loop *loop = ov_os_event_loop(ov_event_loop_config_default());
  ov_sip_app *app = ov_sip_app_create("SIP Test", loop, cfg);

  int exit_value = EXIT_FAILURE;

  if (ov_sip_app_register_handler(app, "BYE", cb_sip_request) &&
      ov_sip_app_register_handler(app, "ACK", cb_sip_request) &&
      ov_sip_app_register_handler(app, "CANCEL", cb_sip_request) &&
      ov_sip_app_register_handler(app, "INVITE", cb_sip_request) &&
      ov_sip_app_register_response_handler(app, cb_sip_response) &&
      ov_sip_app_open_server_socket(app, server_socket_cfg)) {
    print("Listening on %s:%" PRIu16 "\n", server_socket_cfg.host,
          server_socket_cfg.port);
    ov_event_loop_run(loop, OV_RUN_MAX);
    exit_value = EXIT_SUCCESS;

  } else {
    print("Failed to open server socket on %s:%" PRIu16 "\n",
          server_socket_cfg.host, server_socket_cfg.port);
    exit_value = EXIT_FAILURE;
  }

  app = ov_sip_app_free(app);
  loop = ov_event_loop_free(loop);

  constring = ov_free(constring);

  return exit_value;
}
